qiskit/transpiler/passes/optimize_phase_shift_gates.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""Pass for merging/canceling phase-shift gates like u1(theta), T-gates, S-gates and Z-gates
+"""
+import numpy as np
+import qiskit
+from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister
+from qiskit.transpiler.basepasses import TransformationPass
+import copy
+import sys
+np.set_printoptions(threshold=sys.maxsize)
+import warnings
+warnings.filterwarnings('ignore')
+
+class OptimizePhaseShiftGates(TransformationPass):
+    """merge/cancel phase-shift gates in dag."""
+
+    def run(self, dag):
+        """
+        Run one pass of phase-shift gate optimization and cx cancellation
+        on the circuit. The function looks for the circuit configuration
+        with the lowest number of gates and circuit depth
+
+        Args:
+            dag (DAGCircuit): the directed acyclic graph to run on.
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+
+        n = dag.width()  # number of qubits in the circuit
+        T_counter = {}  # Counts the cumulative phase of T-gates and other z-rotation gates
+        T_position_counter = {}  # Keeps track of the locations of T-gates and other z-rot. gates
+        state_tracker = np.zeros((dag.width(), dag.size())).astype('int')
+        for i in range(dag.width()):
+            state_tracker[i][i] = 1 # initial state
+        k = dag.width()  # number of initial variables
+        for e in dag.topological_nodes():
+            if e.data_dict['type'] == 'op':
+                if e.data_dict['name'] == 'cx':
+                    state_tracker[e.data_dict['qargs'][1][1]] ^= state_tracker[e.data_dict['qargs'][0][1]]
+                elif e.data_dict['name'] in ['t', 'tdg', 's', 'sdg', 'z']:
+                    phase_T_gate_equivalent = {'t': 1, 'tdg': 7, 's': 2, 'sdg': 6, 'z': 4}[e.data_dict['name']]
+                    if str(state_tracker[e.data_dict['qargs'][0][1]]) in T_counter:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] += phase_T_gate_equivalent
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])].append(e._node_id)
```

Comment:
Instead of storing the `._node_id` is it possible to store the node itself? The `DAGCircuit` takes `DAGNodes` instead of `node_id`s now so this may clean other areas up too.

Reply:
Excellent idea! I will try to implement this :)

Reply:
I implemented the requested change. Now the nodes are stored, instead of the `._node_id`. Nevertheless, this didn't enable me to completely avoid calling `._node_id` and `.remove_op_node()`, as while optimizing the circuit, it is necessary to make a copy of a circuit, which then contains different nodes (but with the same node id).